packages/react/src/drafts/TabPanels/TabPanels.tsx

```diff
@@ -0,0 +1,160 @@
+import getGlobalFocusStyles from '../../internal/utils/getGlobalFocusStyles'
+
+const TAB_CLASS = 'TabPanel-tab'
+
+const TabContainer = styled(createComponent(TabContainerElement, 'tab-container'))`
+  & > :not([role='tabpanel']) {
+    display: inline-block;
+  }
+
+  &::part(tablist-wrapper) {
+    margin-top: 0;
+    margin-bottom: 16px;
+    border-bottom: 1px solid ${get('colors.border.default')};
+  }
+
+  &:not(:defined) [role='tabpanel'] {
+    margin-top: 17px;
+    display: none;
+  }
+
+  &:not(:defined) [role='tab']:nth-of-type(1)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(1),
+  &:not(:defined) [role='tab']:nth-of-type(2)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(2),
+  &:not(:defined) [role='tab']:nth-of-type(3)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(3),
+  &:not(:defined) [role='tab']:nth-of-type(4)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(4),
+  &:not(:defined) [role='tab']:nth-of-type(5)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(5),
+  &:not(:defined) [role='tab']:nth-of-type(6)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(6),
+  &:not(:defined) [role='tab']:nth-of-type(7)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(7),
+  &:not(:defined) [role='tab']:nth-of-type(8)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(8),
+  &:not(:defined) [role='tab']:nth-of-type(9)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(9),
+  &:not(:defined) [role='tab']:nth-of-type(10)[aria-selected='true'] ~ [role='tabpanel']:nth-of-type(10) {
+    display: block;
+  }
+
+  &:not(:defined):not(:has([aria-selected='true'])) [role='tabpanel']:first-of-type {
+    display: block;
+  }
+
+  &:not(:has([aria-selected='true'])) [role='tab'] ~ [role='tab'] {
+    color: ${get('colors.fg.muted')};
+    background-color: transparent;
+    border: 1px solid transparent;
+  }
+
+  &:not(:has([aria-selected='true'])) [role='tab'],
+  & [role='tab'][aria-selected='true'] {
+    color: ${get('colors.fg.default')};
+    border-color: ${get('colors.border.default')};
+    border-top-right-radius: ${get('radii.2')};
+    border-top-left-radius: ${get('radii.2')};
+    background-color: ${get('colors.canvas.default')};
+  }
+
+  &:not(:defined):not(:has([aria-selected='true'])) [role='tab'] ~ [role='tab'],
+  &:not(:defined):has([aria-selected='true']) [role='tab']:not([aria-selected='true']) {
+    padding: 8px 16px;
+    border-bottom: 1px solid ${get('colors.border.default')};
+  }
+
+  &:not(:defined) :not([role='tabpanel']) {
+    vertical-align: top;
+  }
+
+  ${sx};
+`
+
+export type TabPanelsProps = ComponentProps<typeof TabContainer>
+
+function TabPanels({children, id, ...props}: TabPanelsProps) {
+  // Loop through the chidren, if it's a tab, then add id="{id}-tab-{index}"
+  // If it's a panel, then add aria-labelledby="{id}-tab-{index}"
+  let tabIndex = 0
+  let panelIndex = 0
+
+  const childrenWithProps = React.Children.map(children, child => {
```

Comment:
I think this technique would make it so that people couldn't wrap `Tab` in a custom component as then it wouldn't be able to infer the props from the children needed to setup `id` (or `aria-labelledby` in the panel case). For example:

```tsx
function MyCustomTab({ children }) {
  return <TabPanels.Tab onClick={/* ... */}>{children}</TabPanels.Tab>
}

<TabPanels>
  <MyCustomTab>Tab one</MyCustomTab>
  <MyCustomTab>Tab two</MyCustomTab>
  <MyCustomTab>Tab three</MyCustomTab>
  <TabPanels.Panel>Panel one</TabPanels>
  <TabPanels.Panel>Panel two</TabPanels>
  <TabPanels.Panel>Panel three</TabPanels>
</TabPanels>
```

Reply:
@joshblack can you think of a good way to do this? We could probably do something with `prototype` and `Object.getPrototypeOf` or something like that but I think that's only going to support the case where there's a single parent level. So something like this won't be possible:

```
function MyCustomParentTab({ children }) {
  return <TabPanels.Tab onClick={/* ... */}>{children}</TabPanels.Tab>
}

function MyCustomTab({ children }) {
  return <MyCustomParentTab onChange={/* ... */}>{children}</MyCustomParentTab>
}

<TabPanels>
  <MyCustomTab>Tab one</MyCustomTab>
  <MyCustomTab>Tab two</MyCustomTab>
  <MyCustomTab>Tab three</MyCustomTab>
  <TabPanels.Panel>Panel one</TabPanels>
  <TabPanels.Panel>Panel two</TabPanels>
  <TabPanels.Panel>Panel three</TabPanels>
</TabPanels>
```

Reply:
To be fair, we do already have this limitation with other PRC components that use the `slots` pattern, so I guess it's acceptable. But I agree I've never loved this restriction - it's frustrating because it prevents abstraction.

I'm not sure what a better solution might be in this case though - if you are manipulating `children`, you naturally run into these problems unfortunately. The better way would be to find a solution that avoids manipulating the children [as noted in the API docs](https://react.dev/reference/react/Children) but that's a tricky problem in this case 🤔

Reply:
Totally agreed @iansan5653, it's something that it'd be great to avoid in new components as it makes composition and abstraction limited 😞 

@owenniblock I think one approach could be to use context for this. In situations where people use indexes, the context value may be the index itself. For example:

```tsx
const ItemContext = createContext();

function List({ children }) {
  return (
    <div>
      {Children.map(children, (child, index) => {
        return <ItemContext.Provider value={index}>{child}</ItemContext.Provider>
      })}
    </div>
  );
}
```

This allows for any number of custom components around the item but that item will still know what position it is in. The downside of this, unfortunately, is that things like fragments won't work.

Another strategy that seems common is that items have a unique identifier prop, such as `value` or `id`, that is then used to communicate with a root level context when something changes.

```tsx
const ListContext = createContext();

function List({ children }) {
  return <ListContext.Provider value={/* ... */}>{children}</ListContext.Provider>
}

function Item({ children, value }) {
  const context = useContext(ListContext);

  return <button onClick={() => context.someAction(value)}>{children}</button>
}
```

Hope those examples make sense!

Reply:
IMO, I still think the most straightforward API with the least drawbacks is using a single component for the tabs and panels, like:

```tsx
<TabPanel tab={<><Icon /> Tab name</>}>
  Tab contents
</TabPanel>
```

I [proposed](https://github.com/primer/react/pull/4241#discussion_r1501104692) this earlier and the suggestion was rejected because:

1. Putting JSX in a prop isn't considered a great API
2. It doesn't provide as much flexibility around inserting custom content in to the tab bar

For (1) we can consider the React `Suspense` API which does exactly the same thing with its `fallback` prop. Perhaps it's not beautiful but it's standard and idiomatic. 

(2) is interesting but I think it's worth reconsidering how much customization we really need. Can we just say that any non-tab content is rendered in the tab bar where it goes? 

```tsx
Content before panel
<TabPanels>
  Content before tabs in tab bar
  <TabPanel tab={<><Icon /> Tab name</>}>
    Tab contents
  </TabPanel>
  Content between tabs in tab bar
  <TabPanel tab={<><Icon /> Tab name</>}>
    Tab contents
  </TabPanel>
  Content after tabs in tab bar
</TabPanels>
Content after panel
```

Is there anything that this approach really doesn't support?

Reply:
I wonder if these discussions should happen after we release this as a draft and before we move to alpha? 🤔 I'm aware that it's taken a long time to get us this far and our priority should be getting the `CommentEditor` experience out and in line with the Rails experience.

If we feel like we need to get the API hammered out before we release this as a PRC component, perhaps an alternative approach could be to move this work into dotcom as an experimental component and then iterate there before moving it back to Primer later?

Happy with either approach, just don't want to spend too much longer this bit of work and lose sight of our goals.

If neither of those approaches is acceptable, I guess I just need a clear idea of what our definition of done is here so I know what I need to finish off to get this merged.

Reply:
In the meantime @joshblack I've changed this logic to branch on `child.props.role` instead of the type of the child, hopefully this will enable consumers to create wrapper elements for the tabs and tabpanels.

Reply:
> I guess I just need a clear idea of what our definition of done is here so I know what I need to finish off to get this merged.

Yeah I should clarify that for me personally, while I have strong opinions about React APIs, my opinions are definitely not blocking. Feel free to disregard my comments if you have a need to get this shipped 🙂 



> In the meantime @joshblack I've changed this logic to branch on `child.props.role` instead of the type of the child, hopefully this will enable consumers to create wrapper elements for the tabs and tabpanels.

I don't think this will actually solve the problem unfortunately. As a basic example:

```tsx
const PreviewTab = ({markdown}) => {
  const content = useRenderMarkdown(markdown)

  return <TabPanel>{content}</TabPanel>
}


const CommentEditor = ({value}) => (
  <TabPanels>
    <TabPanels.Tab>Preview</TabPanels.Tab>
    <PreviewTab markdown={value} />
  </TabPanels>
)
```

In this example, when iterating over the children you'll see a `PreviewTab` instance. `child.props` will be an object in the shape `{markdown: string}` - there's no `role` here. 

If we have to do this manipulation, a better approach might be to directly manipulate the DOM instead of working with the `Children` API. For example, you could, in a `useEffect` after rendering, use `querySelectorAll("[role=tab]")` to get all the tab elements, then update them. This allows any components to be used since it only depends on the DOM structure.

However it's worth noting that this is just as much of a hack as using the `Children` API and you have to be very careful to keep reapplying the changes as React updates the DOM. It's also not SSR-compatible; the manipulation has to be done after client-side render. It's very much not ideal but it is a lot more flexible than manipulating `children`.

The 'idiomatic' way to do it would still be the prop slot API I proposed in my previous comments, since that requires no manipulation outside the standard React render cycle.